third_party/move/move-compiler-v2/src/ast_simplifier.rs

```diff
@@ -0,0 +1,1187 @@
+                        use ExpData::*;
+                        match rexp.as_ref() {
+                            LocalVar(id, ..)
+                            | Temporary(id, ..)
+                            | Call(id, Operation::Select(..), _) => {
+                                let cloned_id = self.env().clone_node(*id);
+                                Sequence(cloned_id, vec![rexp]).into_exp()
+                            },
+                            _ => {
+                                // Nothing to do.
+                                rexp
+                            },
+                        }
+                    } else {
+                        rexp
+                    }
+                } else {
+                    rexp
+                }
+            },
+        };
+        let protected_rexp_id = protected_rexp.as_ref().node_id().as_usize();
+        trace!(
+            "After rewrite2, {} is now protected_rexp {}: `{}`",
+            new_id,
+            protected_rexp_id,
+            protected_rexp.display_verbose(self.env())
+        );
+        protected_rexp
+    }
+
+    fn rewrite_enter_scope<'a>(
+        &mut self,
+        _id: NodeId,
+        vars: impl Iterator<Item = &'a (NodeId, Symbol)>,
+    ) {
+        self.visiting_binding.enter_scope();
+        self.values.enter_scope();
+        for (_, sym) in vars {
+            self.values.remove(*sym);
+        }
+    }
+
+    fn rewrite_exit_scope(&mut self, _id: NodeId) {
+        self.visiting_binding.exit_scope();
+        self.values.exit_scope();
+    }
+
+    fn rewrite_local_var(&mut self, id: NodeId, sym: Symbol) -> Option<Exp> {
+        // Note that we could but don't need to check `in_mut_borrow` since if we have a value for a
+        // variable here then the variable can't appear in a borrow position of a Mutable borrow
+        // (see `find_possibly_modified_vars`).
+        let result = self.rewrite_to_recorded_value(id, &sym);
+        if log_enabled!(Level::Trace) {
+            if let Some(exp) = &result {
+                let in_scope = self.visiting_binding.get(&sym);
+                let value = self.values.get(&sym);
+                trace!(
+                    "Replacing symbol {} at node {} with {}; in_scope={:?}, value={:?}",
+                    sym.display(self.env().symbol_pool()),
+                    id.as_usize(),
+                    exp.display(self.env()),
+                    in_scope.map(|n| n.as_usize()),
+                    value
+                );
+                assert!(!self.in_mut_borrow);
+            }
+        }
+        result
+    }
+
+    fn rewrite_call(&mut self, id: NodeId, oper: &Operation, args: &[Exp]) -> Option<Exp> {
+        self.constant_folder
+            .rewrite_call(id, oper, args)
+            .map(|exp| self.expand_tuple(exp))
+            .or_else(|| {
+                // Not completely a constant.
+                if *oper == Operation::Cast && args.len() == 1 {
```

Comment:
If oper is `cast` (unary operator), `args.len()` has to be 1?

maybe assert if you don't trust this invariant?

Reply:
This is an optimization, so I prefer not to do anything noisy here.